flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.changelog;
+
+import org.apache.flink.api.common.state.State;
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.state.changelog.StateChange;
+import org.apache.flink.runtime.state.internal.InternalKvState;
+import org.apache.flink.runtime.state.internal.InternalValueState;
+
+import java.io.IOException;
+
+/**
+ * Delegated partitioned {@link ValueState} that forwards changes to {@link StateChange} upon {@link
+ * ValueState} is updated.
+ *
+ * @param <K> The type of the key.
+ * @param <N> The type of the namespace.
+ * @param <V> The type of the value.
+ */
+class ChangelogValueState<K, N, V> implements InternalValueState<K, N, V> {
```

Comment:
We could introduce `AbstractChangelogState` to act as `InternalKvState`, so that we could only keep only one place of those `getKeySerializer`, `getNamespaceSerializer`, `getSerializedValue` and so on.

Maybe you could refer to [pull/15119](https://github.com/apache/flink/pull/15119/files#diff-52de1f1a9a4e344e62306261f83533bc6a87326f372d775bc984c3a944d66943) for similar implementation.

Reply:
Thanks, I've tried this way, and I like it a lot better.